### PR TITLE
EvaluateRubricJob calls openai via aiproxy

### PR DESCRIPTION
**Pull request by davidsbailey**
Finishes https://codedotorg.atlassian.net/browse/AITT-243 (updated link) by connecting the EvaluateRubricJob to the python service to evaluate the rubric. New in this PR:
* fetch the ai rubric, prompt, and examples from S3
* send the rubric, prompt and examples to openai via the aiproxy (see https://github.com/code-dot-org/aiproxy/tree/openai-routes for how to run locally)

as a side note, getting examples working required this code change on the python side: https://github.com/code-dot-org/aiproxy/commit/1be7f4c12e553c8fab356e233d3d837addf03917

## Testing story

* manual testing end to end with locally running ai proxy
* add unit tests stubbing out s3  using stub_responses (see [docs](https://aws.amazon.com/blogs/developer/advanced-client-stubbing-in-the-aws-sdk-for-ruby-version-3/)) and by stubbing out the network call to openai

### CODE REVIEW

File: dashboard/app/jobs/evaluate_rubric_job.rb [hunk 1]
```diff
@@ -79,12 +90,58 @@ def self.get_lesson_s3_name(script_level)
...
+  private def read_file_from_s3(lesson_s3_name, key_suffix)
+    bucket = 'cdo-ai'
+    key = "teaching_assistant/lessons/#{lesson_s3_name}/#{key_suffix}"
+    s3_client.get_object(bucket: bucket, key: key)[:body].read
+  end
+
+  private def read_examples(lesson_s3_name)
+    bucket = 'cdo-ai'
+    prefix = "teaching_assistant/lessons/#{lesson_s3_name}/examples/"
+    response = s3_client.list_objects_v2(bucket: bucket, prefix: prefix)
```
Comment (wilkie) [inline dashboard/app/jobs/evaluate_rubric_job.rb]: I wonder if it makes sense to consider caching these files where the job worker is somehow, than to negotiate and download these files for every student submission when they are not likely to change. (Would a memcache/redis service be available, and if so, could it be used for that)

Comment (wilkie) [reply dashboard/app/jobs/evaluate_rubric_job.rb]: Perhaps we can revisit this more later.

File: dashboard/app/jobs/evaluate_rubric_job.rb [hunk 2]
```diff
@@ -79,12 +90,58 @@ def self.get_lesson_s3_name(script_level)
...
-        'Grade' => understanding_s
-      }
+  private def read_file_from_s3(lesson_s3_name, key_suffix)
+    bucket = 'cdo-ai'
+    key = "teaching_assistant/lessons/#{lesson_s3_name}/#{key_suffix}"
+    s3_client.get_object(bucket: bucket, key: key)[:body].read
+  end
+
+  private def read_examples(lesson_s3_name)
+    bucket = 'cdo-ai'
```
Comment (wilkie) [inline dashboard/app/jobs/evaluate_rubric_job.rb]: This is hard coded in a few places. Could we make the bucket name a constant? We have this similar pattern in a few places: `S3_AI_BUCKET = 'cdo-ai'.freeze`

**Review by wilkie — APPROVED** [review]
I only have a few nits. Looks good. I see you've made the changes to match the changes I made to the proxy service, already. Ugh the way `AWS` and `Aws` are different but similar things gave me a slight headache, but alas.

File: dashboard/app/jobs/evaluate_rubric_job.rb (hunk 1 above)
Comment (davidsbailey) [reply dashboard/app/jobs/evaluate_rubric_job.rb]: yeah, I agree that we should do this, but also think we can get away without it for the pilot. tracking it here along with a link to one way of doing it: https://codedotorg.atlassian.net/browse/AITT-250

File: dashboard/app/jobs/evaluate_rubric_job.rb (hunk 2 above)
Comment (davidsbailey) [reply dashboard/app/jobs/evaluate_rubric_job.rb]: good idea! done

**Comment by davidsbailey** [comment]
> I only have a few nits. Looks good. I see you've made the changes to match the changes I made to the proxy service, already. 

yes, which reminds me that we have some integration testing ahead of us. I don't think a testing environment is explicitly part of our request to the infra team, so it may be that we are going to have to do extra work to get automated integration testing working (currently tracked here but without many details: https://codedotorg.atlassian.net/browse/AITT-133)

> Ugh the way `AWS` and `Aws` are different but similar things gave me a slight headache, but alas.

yes, this is super annoying! It might actually have been simpler to keep the same name, even though it would mean that we would be monkey-patching. a more different name would probably have been the best choice.

**Pull request merged**